### PR TITLE
[CIR] Add emitDeclInvariant for global with constant storage

### PR DESCRIPTION
**Pull request by adams381**
Implement emitDeclInvariant to emit llvm.invariant.start intrinsic for global variables with constant storage. This enables optimizations by marking when a global becomes read-only after initialization.

## Changes
- Add emitDeclInvariant and emitInvariantStart functions in CIRGenCXX.cpp
- Add emitInvariantStart declaration in CIRGenFunction.h
- Update emitCXXGlobalVarDeclInit to call emitDeclInvariant for constant storage globals after initialization
- Update getOrCreateCIRGlobal to set constant flag on globals with constant storage
- Add comprehensive test covering positive and negative cases

## Implementation Details
The implementation handles address spaces correctly, dynamically constructing the intrinsic name (e.g., invariant.start.p0, invariant.start.p10) based on the pointer's address space. The intrinsic is only emitted when optimizations are enabled, matching classic codegen behavior.

## Testing
All tests pass (411/412, 1 unsupported). The test file includes CIR, LLVM, and OGCG checks for both optimized and non-optimized builds.

### CODE REVIEW

File: clang/lib/CIR/CodeGen/CIRGenModule.cpp [hunk 1]
```diff
@@ -652,10 +652,19 @@ CIRGenModule::getOrCreateCIRGlobal(StringRef mangledName, mlir::Type ty,
 
   mlir::Location loc = getLoc(d->getSourceRange());
 
+  // Calculate constant storage flag before creating the global.
```
Comment (andykaylor) [inline clang/lib/CIR/CodeGen/CIRGenModule.cpp]: Why did you move this?

File: clang/lib/CIR/CodeGen/CIRGenModule.cpp [hunk 2]
```diff
@@ -870,10 +875,14 @@ void CIRGenModule::emitGlobalVarDefinition(const clang::VarDecl *vd,
     emitter->finalize(gv);
 
   // If it is safe to mark the global 'constant', do so now.
+  // Use the same logic as emitCXXGlobalVarDeclInit to determine constant
```
Comment (andykaylor) [inline clang/lib/CIR/CodeGen/CIRGenModule.cpp]: This code does not use the same logic as emitCXXGlobalVarDeclInit in classic codegen. I don't know why it doesn't but it doesn't. This change could lead us to behave differently.

File: clang/test/CIR/CodeGen/global-constant-storage.cpp [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+public:
+  C();
+  int a;
+  int b;
+};
+
+const C c;
+
+// CIR checks for 'a' - should have constant storage
+// CIR: cir.global constant external @a = #cir.zero : !rec_A
```
Comment (andykaylor) [inline clang/test/CIR/CodeGen/global-constant-storage.cpp]: The initial CIR generated for this case should look like this:

```
  cir.global external dso_local @a = ctor : !rec_A {
    %0 = cir.get_global @a : !cir.ptr<!rec_A>
    cir.call @_ZN1AC1Ev(%0) : (!cir.ptr<!rec_A>) -> ()
  }
```
That gets transformed into what you're checking here during LoweringPrepare, but I'd like to see checks for both states. You can do that by passing `-mmlir --mlir-print-ir-before=cir-lowering-prepare` on the RUN line and capturing the stderr output to a different file. Look at `clang/test/CIR/CodeGen/global-init.cpp` for an example.

**Comment by adams381** [comment]
All reviewer feedback has been addressed:

✅ Parameter naming - Changed to camelCase
✅ Address space handling - Removed dynamic suffix construction
✅ Comment explaining move - Added explanation for constant storage calculation move
✅ Logic consistency - Aligned emitGlobalVarDefinition with classic codegen
✅ Remove CWithDtor case - Removed duplicate test case
✅ Reorganize test checks - Interleaved checks by test case
✅ Add checks for before/after LoweringPrepare - Added CIR-BEFORE-LPP checks
✅ Add -O1 RUN line for CIR output - Added CIR-O1 checks

Ready for re-review.

**Review by andykaylor — APPROVED** [review]
lgtm
